[Win] Enable WebDriver automation with wincairo as client

haruhisa-shin opened the pull request:
#### db706f64b58197fa0c3fc1dda5303aecc8d5fdf8
<pre>
[Win] Enable WebDriver automation with wincairo as client
<a href="https://bugs.webkit.org/show_bug.cgi?id=198055">https://bugs.webkit.org/show_bug.cgi?id=198055</a>

Reviewed by Fujii Hironori.

Allows UIProcess of wincairo to be controlled by WebDriver.

Page transitions, manipulation of elements, and keyboard/mouse
insteractions are possible.
However, some functions are not yet supported.

Also, until now, automation has been started on MainThread
and terminated on WorkerThread of the socket.
To avoid assertion fail by canSafelyBeUsed() of WeakRef,
changed the termination process to be performed on MainThread.

How to use:
1. launch MiniBrowser.exe with environment variable WEBKIT_INSPECTOR_SERVER.
&gt; set WEBKIT_INSPECTOR_SERVER=127.0.0.1:1641
&gt; WebKitBuild\Release\bin64\MiniBrowser.exe

2. launch WebDriver.exe on another console.
&gt; WebKitBuild\Release\bin64\WebDriver.exe -t 127.0.0.1:1641 -p 12345

Then you would be able to use WebDriver by sending commands to
url &quot;<a href="http://127.0.0.1">http://127.0.0.1</a>:12345&quot;.

If you use Selenium, you could navigate your browser
with a script like the following:

```
from selenium import webdriver

options = webdriver.WebKitGTKOptions()
driver = webdriver.Remote(
    command_executor=&quot;<a href="http://{0}">http://{0}</a>:{1}&quot;.format(&apos;127.0.0.1&apos;, 12345),
    options=options
)

try:
    driver.get(&apos;<a href="https://docs.webkit.org/Ports/WindowsPort.html&apos">https://docs.webkit.org/Ports/WindowsPort.html&apos</a>;)
finally:
    driver.quit()
```

* Source/WebDriver/socket/SessionHostSocket.cpp:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
* Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp: Added.
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/win/AutomationClientWin.cpp: Added.
* Source/WebKit/UIProcess/win/AutomationClientWin.h: Added.
* Source/WebKit/UIProcess/win/WebProcessPoolWin.cpp:
* Source/cmake/OptionsWin.cmake:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: <a href="https://commits.webkit.org/274530@main">https://commits.webkit.org/274530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9ea8e472dfc3ef0533d2bd3e36335f22167dcba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18359 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41734 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41914 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35280 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41686 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21226 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15688 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39954 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34108 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13427 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35050 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43192 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/32833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35746 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35381 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39194 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39006 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/11699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15798 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46012 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/15846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9373 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5160 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -0,0 +1,107 @@
+        auto& processProxy = processPool->processes()[0];
+        if (processProxy->pageCount()) {
+            found = true;
+            completionHandler(&((*(processProxy->pages().begin())).get()));
```
I think you can use `processProxy[0]` to get the first item. And, you can use `ptr()` instead of `get()` here.

fujii commented on Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp:
```diff
@@ -0,0 +1,365 @@
+    String keyIdentifier = "Key"_s + keySequence.convertToASCIIUppercase();
+
+    auto hwnd = reinterpret_cast<HWND>(page.viewWidget());
+    String key = (m_currentModifiers & kWKEventModifiersShiftKey)? keySequence.convertToASCIIUppercase(): keySequence;
```
Not documented, but I think you have to add white spaces before `:` and `?` in WebKit..  

`x ? a : b`, not `x? a: b`.

aprotyas replied:
Documented here! https://webkit.org/code-style-guidelines/#spacing-binary-ternary-op

haruhisa-shin replied:
I missed it because the `Tools/Scripts/check-webkit-style` did not make an error.
Thanks for pointing this out.

fujii commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -365,6 +365,11 @@ Inspector::Protocol::ErrorStringOr<void> WebAutomationSession::closeBrowsingCont
+#if PLATFORM(WIN) // TODO: also PLATFORM(PLAYSTATION) should be added
+    // We don't have to close page/window/application, but want to disconnect to automation frontend.
+    terminate();
+#endif
```
Why don't we have to close? If you don't have to close, can we put it in `#else`?
~~~
#else
  page->closePage();
#endif
~~~

haruhisa-shin replied:
> Why don't we have to close? 

The wincairo WebDriver enabled by this change does not launch a browser when connecting.
WebDriver connects to a browser that is already running and performs automation.
So, I thought it would be better not to close the browser even when disconnecting.

> If you don't have to close, can we put it in #else?

You are right, `closePage()` is not needed and can be placed in #else.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -0,0 +1,107 @@
+    if (processPool && processPool->processes().size()) {
+        // Set controlledByAutomation flag to all pages in all processes
+        processPool->setPagesControlledByAutomation(true);
+        // The default page for automation target is the first page of the first process
```
Why not openning a new window?

haruhisa-shin replied:
The purpose of this patch is to enable the same functionality as the playstation port, and because I adopt the approach of connecting to the webview remotely.
(This is the same information I explained in the slack the other day)
As you pointed out, I'll add "how to use" to the commit log, so please refer to it as well.

fujii commented on Source/WebKit/UIProcess/win/WebProcessPoolWin.cpp:
```diff
@@ -86,4 +89,14 @@ void WebProcessPool::platformResolvePathsForSandboxExtensions()
 }
 
+#if ENABLE(REMOTE_INSPECTOR)
+void WebProcessPool::setPagesControlledByAutomation(bool controlled)
```
Do you have a plan to use this method for PlayStation port too?
If so, why not adding this to the common WebProcessPool.cpp with `#if PLATFORM(WIN)`?

haruhisa-shin replied:
Yes, we will use this method for playstation port.
I'll move this to the common WebProcessPool.cpp.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.h:
```diff
@@ -0,0 +1,78 @@
+    void requestAutomationSession(const String&, const Inspector::RemoteInspector::Client::SessionCapabilities&) override;
+    void closeAutomationSession() override;
+
+    WebProcessPool& m_processPool;
```
This should be a smart pointer rather than a raw reference.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines

haruhisa-shin replied:
changed to `WeakPtr<WebProcessPool>`.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -0,0 +1,111 @@
+void AutomationClient::closeAutomationSession()
+{
+    RunLoop::main().dispatch([this] {
+        if (!m_processPool->automationSession())
```
I think you have to check the WeakPtr before using it.
	   
~~~
  if (!m_processPool)
    return;
~~~

fujii replied:
You have to keep a RefPtr of m_processPool to call its methods.
https://github.com/WebKit/WebKit/wiki/Smart-Pointer-Usage-Guidelines
> RefPtr processPool = m_processPool;
> if (!processPool)
>     return;

fujii replied:
I think it might be a good idea to add a new getter `protectedProcessPool()`.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -0,0 +1,111 @@
+
+void AutomationClient::closeAutomationSession()
+{
+    RunLoop::main().dispatch([this] {
```
I think you need to retain a ref count of `this` in this lambda capture.

haruhisa-shin committed:
[Win] Enable WebDriver automation with wincairo as client
https://bugs.webkit.org/show_bug.cgi?id=198055

Reviewed by Fujii Hironori.

Allows UIProcess of wincairo to be controlled by WebDriver.

Page transitions, manipulation of elements, and keyboard/mouse
insteractions are possible.
However, some functions are not yet supported.

Also, until now, automation has been started on MainThread
and terminated on WorkerThread of the socket.
To avoid assertion fail by canSafelyBeUsed() of WeakRef,
changed the termination process to be performed on MainThread.

How to use:
1. launch MiniBrowser.exe with environment variable WEBKIT_INSPECTOR_SERVER.
> set WEBKIT_INSPECTOR_SERVER=127.0.0.1:1641
> WebKitBuild\Release\bin64\MiniBrowser.exe

2. launch WebDriver.exe on another console.
> WebKitBuild\Release\bin64\WebDriver.exe -t 127.0.0.1:1641 -p 12345

Then you would be able to use WebDriver by sending commands to
url "http://127.0.0.1:12345".

If you use Selenium, you could navigate your browser
with a script like the following:

```
from selenium import webdriver

options = webdriver.WebKitGTKOptions()
driver = webdriver.Remote(
    command_executor="http://{0}:{1}".format('127.0.0.1', 12345),
    options=options
)

try:
    driver.get('https://docs.webkit.org/Ports/WindowsPort.html')
finally:
    driver.quit()
```

* Source/WebDriver/socket/SessionHostSocket.cpp:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
* Source/WebKit/UIProcess/Automation/win/WebAutomationSessionWin.cpp: Added.
* Source/WebKit/UIProcess/WebProcessPool.cpp:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/win/AutomationClientWin.cpp: Added.
* Source/WebKit/UIProcess/win/AutomationClientWin.h: Added.
* Source/WebKit/UIProcess/win/WebProcessPoolWin.cpp:
* Source/cmake/OptionsWin.cmake:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: https://commits.webkit.org/274530@main
